File: 0125-valid-palindrome/0125-valid-palindrome.java
class Solution {
    public boolean isPalindrome(String s) {
        // Clean up the string: ignore non-alphanumeric characters and make it lowercase
        s = s.toLowerCase().replaceAll("[^a-z0-9]", "");
        
        // Call the recursive helper function
        return checkpal(s, 0, s.length() - 1);
    }
    
    private boolean checkpal(String s, int start, int end) {
        // Base case: if the start index crosses the end index, it's a palindrome
        if (start >= end) return true;
        
        // Recursive case: check if characters at start and end are the same
        if (s.charAt(start) != s.charAt(end)) return false;
        
        // Recurse for the next characters
        return checkpal(s, start + 1, end - 1);
    }
}

File: 0125-valid-palindrome/0125-valid-palindrome.py
class Solution:
    def isPalindrome(self, s: str) -> bool:
        # Clean up the string: ignore non-alphanumeric characters and make it lowercase
        s = ''.join(char.lower() for char in s if char.isalnum())
        
        # Call the recursive helper function
        return self.checkpal(s, 0, len(s) - 1)
    
    def checkpal(self, s: str, start: int, end: int) -> bool:
        # Base case: if the start index crosses the end index, it's a palindrome
        if start >= end:
            return True
        
        # Recursive case: check if characters at start and end are the same
        if s[start] != s[end]:
            return False
        
        # Recurse for the next characters
        return self.checkpal(s, start + 1, end - 1)
